### ISSUE TITLE
IDE0300 false-positive

### ISSUE THREAD

**Issue by meziantou** [dotnet/roslyn#70816]
**Version Used**: 4.9.0-1.23560.3 (78d23490)

**Steps to Reproduce**:

- Use .NET 8 (maybe it can be reproduced using previous version of .NET)
- `dotnet new console`
- Replace `Program.cs` with the following content:

````c#
byte[] a = [1, 2];
a.AsSpan().SequenceEqual(new byte[] { 0, 1 }); // IDE0300
````

If you apply the IDE0300 fix

````
byte[] a = [1, 2];
a.AsSpan().SequenceEqual([0, 1]);

// CS1929 'Span<byte>' does not contain a definition for 'SequenceEqual' and the best extension method overload 'MemoryExtensions.SequenceEqual<int>(ReadOnlySpan<int>, ReadOnlySpan<int>)' requires a receiver of type 'System.ReadOnlySpan<int>'
````

https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKAOAE8AXAUwG0BdAAgloF5bLs5bNqBuAiAOgCCAZwDKABwgwAFAEo+I8gEcAruRgBjcgFEVEADZSY5AO60SFGrQDetAAztstAL4yutAPTvaASQAiW2wBmW1teQVEJaTkFFTVNHWV9KUp7WmxqVw8vXwB7ciEYAHJSWnUcsDEoPXJ6IVooEtIAJ2JaUhzaADNYBFoAJXIIBAB5GD1icUkAHlhSAD4CIA===

**Diagnostic Id**: IDE0300

**Expected Behavior**:

No diagnostic as the fix is not applicable

**Actual Behavior**:

It reports a diagnostic

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson this seems odd. Why is the span overload not applicable here?

**Comment by ufcpp** [comment]
Not sure if it is the same cause, but the following code also has false positives.

```cs
object falsePositive = new[] { new[] { 1 }, new[] { 1 } }; // IDE0300

object forComparison = new[] { new[] { 1 } }; // No IDE0300
```

**Comment by CyrusNajmabadi** [comment]
@ufcpp that;s unrelated.  can you open a new issue?

**Comment by ufcpp** [comment]
@CyrusNajmabadi
Sorry, it might be fixed in 17.9p1.
![image](https://github.com/dotnet/roslyn/assets/3760955/de82c1a0-dd66-414c-aef0-1480711fa82a)

**Comment by CyrusNajmabadi** [comment]
@ufcpp I also can't repro it

**Comment by RikkiGibson** [comment]
> @RikkiGibson this seems odd. Why is the span overload not applicable here?

Not sure yet. I can see casting the argument to `Span<byte>` makes things start working. It might be something in overload resolution. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKAOAE8AXAUwG0BdAAgloF5bLs5bNqBuAiAOgCCAZwDKABwgwAFAEo+I8gEcAruRgBjcgFEVEADZSY5AO60SFGrQDetAAztstAL4yutAPTvaASQAiW2wBmW1teQVEJaTkFFTVNHWV9KSlxSQAec3IAPhlKe1psalcPLwB7AGsw4VSo+SVVDW1dAzyHIrdPWl9S8iEYAHJSWnVSsDEoPXJ6IVooIdIAJ2JaUlLaADNYBFoAJXIIBAB5GD1iGrTYUiyCIA===)

@cston for visibility.

**Comment by jgilbert2017** [comment]
also, please publish the documentation for the new code analysis codes.  the url below does not exist
https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0300

**Comment by vankampenp** [comment]
I think it is more general. The following also generates false positives:
return new[] { "a", "b" };
with fix becomes
 return ["a", "b"];
This gives a range of errors with red underline
; expected
code is unreachable, identifier expected

**Comment by CyrusNajmabadi** [comment]
@vankampenp Can you supply a repro.  That conversion is correct and expected.

**Comment by vankampenp** [comment]
[ConsoleApp1.zip](https://github.com/dotnet/roslyn/files/13452617/ConsoleApp1.zip)
![error screenshot](https://github.com/dotnet/roslyn/assets/16002371/f05f691b-849e-49a5-957b-953dd1f26269)

**Comment by CyrusNajmabadi** [comment]
Compiles and builds just fine:

![image](https://github.com/dotnet/roslyn/assets/4564579/61e5c84a-943d-4af1-b67e-23c3ddeb8a46)

**Comment by CyrusNajmabadi** [comment]
@vankampenp what version of visual studio are you using?

**Comment by vankampenp** [comment]
Version 17.8.1
I see the problem is Resharper. Sorry for the inconvenience.

**Comment by CyrusNajmabadi** [comment]
> I see the problem is Resharper. Sorry for the inconvenience.

No worries.  Have a good day :)

**Comment by cston** [comment]
For the expression `a.AsSpan().SequenceEqual([0, 1])`, it looks like the relevant overloads for `SequenceEqual()` are from `System.MemoryExtensions`:

```csharp
public static bool SequenceEqual<T>(this Span<T> span, ReadOnlySpan<T> other) where T : IEquatable<T>?;

public static bool SequenceEqual<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> other) where T : IEquatable<T>?;
```

For the first overload, method type inference *fails* because the first argument of type `Span<byte>` contributes an exact bounds of `byte`, and the second argument `[0, 1]` contributes lower bounds of `int`, and there is no implicit conversion from `int` to `byte`.

For the second overload, method type inference *succeeds* because the first argument of type `Span<byte>` does not contribute to type inference, and the second argument `[0, 1]` contributes lower bounds of `int`, so the inferred method is `SequenceEqual<int>(this ReadOnlySpan<int> span, ReadOnlySpan<int> other)`.

Overload resolution then *fails* for `SequenceEqual<int>(this ReadOnlySpan<int> span, ReadOnlySpan<int> other)` since there is no conversion for the first argument from `Span<byte>` to `ReadOnlySpan<int>`.

**Comment by CyrusNajmabadi** [comment]
> For the first overload, method type inference fails because the first argument of type Span<byte> contributes an exact bounds of byte, and the second argument [0, 1] contributes lower bounds of int, and there is no implicit conversion from int to byte.

Something seems wonky here.  `[0,1]` is certainly target typeable to `ReadOnlySpan<byte>`.  I get that there are bounds here of int, but i guess i don't know why it is *lower* bounds as 0/1 are certainly totally reasonable being in the byte-space.

Basically, something seems wonky here :)

**Comment by cston** [comment]
> I get that there are bounds here of int, but i guess i don't know why it is lower bounds as 0/1 are certainly totally reasonable being in the byte-space.

This matches the type inference behavior when the elements are passed as arguments directly (see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFDIMoAcCGA7APAIwE8AXAUwD4ACAZy20oF5LtSB3SosgbQF1KBvSnAA0lGJQC+AbmRpSARwCupbAGNSAUSWYANgApaOUQAZRcAJQyUSKHABsHAPaOdlVAuVrN2nbgAq5HoYOP5UhtiifpQAHpGUhOaMVABmutSkVshAA)):
```csharp
Span<byte> span = new byte[] { 1, 2 };

SequenceEqual(span, 0, 1);

static bool SequenceEqual<T>(Span<T> span, T x, T y) => false;
```
```
error CS0411: The type arguments for method 'SequenceEqual<T>(Span<T>, T, T)' cannot be inferred from the usage
```

**Comment by CyrusNajmabadi** [comment]
Interesting.  Seems like something unfortunate about inference then.  @MadsTorgersen ?

**Comment by RikkiGibson** [comment]
I think in order to fix this, we would have to adjust type argument inference to allow numeric constants to contribute their values to the inference. It would be fairly tricky to get right.

**Comment by CyrusNajmabadi** [comment]
Ok.  So i'll work on the IDE side to fix our analyzer.  This is an interesting thing to go to discussion on csharplang about better numeric inference in general.

> It would be fairly tricky to get right.

My faith in you figuring this out is boundless.

**Comment by CyrusNajmabadi** [comment]
Actually, i don't repro this in latest roslyn.  Will write up a test to validate.
